File: solutions/julia/nucleotide-count/3/nucleotide-count.jl
"""
    count_nucleotides(strand)

The count of each nucleotide within `strand` as a dictionary.

Invalid strands raise a `DomainError`.

"""
function count_nucleotides(strand)
    all([nucleotid in "ACGT" for nucleotid in strand]) || 
         throw(DomainError(strand, "Invalid DNA sequence"))
    merge(Dict('A'=>0, 'C'=>0, 'G'=>0, 'T'=>0),
        Dict(nucleotid => count( ==(nucleotid), strand) for nucleotid in strand))
end

File: solutions/csharp/nucleotide-count/1/NucleotideCount.cs
public static class NucleotideCount
{
    public static IDictionary<char, int> Count(string sequence)
    { 
        var result =  new Dictionary<char, int> 
        {
            ['A'] = 0,
            ['C'] = 0,
            ['G'] = 0,
            ['T'] = 0            
        };
        foreach(char n in sequence){
            if (result.ContainsKey(n)){
                result[n] += 1;
            }
            else {
                throw new ArgumentException("Invalid nucleotid.");
            }
        }
        return result;
    }
}